packages/agents/src/agents/Chameleon/ContextualizedChat.ts

```diff
@@ -0,0 +1,294 @@
+
+    // If we've already processed all messages
+    if (lastProcessedIdx === messages.length - 1) {
+      return Promise.resolve();
+    }
+
+    // Process all messages from lastProcessedIdx -> messages.length
+    for (let i = lastProcessedIdx + 1; i < messages.length; ++i) {
+      const message = messages[i];
+      await this._processNewMessage(message, i, type);
+    }
+  }
+
+  private async _processNewMessage(message: ChatMessage, msgPtr: number, type: ChatLogType): Promise<void> {
+
+    const chunks: string[] = [];
+
+    // If the message is large and requires chunking
+    if (this._chunker.shouldChunk(message)) { 
+      chunks.push(...this._chunker.chunk(message));
+    } else {
+      // TODO: should this be what is used for embedding?
+      //       should we instead just extract the values of the struct?
+      chunks.push(JSON.stringify(message));
+    }
+
+    const chunkMetadata = this._chunkMetadata[type];
+
+    // Save the starting chunk index (used later for document index)
+    const startChunkIdx = chunkMetadata.length;
+
+    // Add message index pointers for each new chunk
+    for (const _ of chunks) {
+      chunkMetadata.push({
+        msgPtr
+      });
+    }
+
+    // If we're processing persistent messages,
+    // and it isn't a large message, early out
+    if (type === "persistent" && chunks.length === 1) {
+      return Promise.resolve();
+    }
+
+    // Create an array of chunk document metadata
+    const metadatas: DocumentMetadata[] = chunks.map((chunk, index) => ({
+      chunkMetadataPtr: startChunkIdx + index,
+      tokens: this._chat.tokenizer.encode(chunk).length
+    }));
+
+    // Add the chunks to the vectordb collection
+    await this._collections[type].add(chunks, metadatas);
+
+    return Promise.resolve();
+  }
+
+  private _getLastProcessedMessageIndex(type: ChatLogType): number {
+    const chunkMsgPtrs = this._chunkMetadata[type];
+    const lastIdx = chunkMsgPtrs.length - 1;
+
+    // Nothing has been processed
+    if (lastIdx < 0) {
+      return -1;
+    }
+
+    // Return the message index of the last metadata
+    return chunkMsgPtrs[lastIdx].msgPtr;
+  }
+
+  private _getSmallMsgs(type: ChatLogType): ChunkMetadataPtr[] {
+    const chunkMetadataPtrs: ChunkMetadataPtr[] = [];
+    const chunkMetadata = this._chunkMetadata[type];
+    let prevMsgPtr = -1;
+
+    // Iterate through and grab all unique msg ptrs.
+    // If a msg is not small, it will be chunked,
+    // meaning more than one ptr will exist
+    for (let i = 0; i < chunkMetadata.length; ++i) {
+      const currMsgPtr = chunkMetadata[i].msgPtr;
+      const nextMsgPtr = i + 1 < chunkMetadata.length ? chunkMetadata[i + 1] : -1;
+      if (currMsgPtr !== prevMsgPtr && currMsgPtr !== nextMsgPtr) {
+        // We've found a unique ptr with no siblings, it is small
+        chunkMetadataPtrs.push(i);
+      }
+    }
+
+    return chunkMetadataPtrs;
+  }
```

Comment:
This implementation would be far simpler if this method was immutable

Reply:
okay nice, will do